pyccel/parser/semantic.py

```diff
@@ -3785,212 +3911,263 @@ def _visit_FunctionDef(self, expr):
+                    else:
+                        result_pointer_map[r] = persistent_targets
+
+            optional_inits = []
+            for a in arguments:
+                var = self._optional_params.pop(a.var, None)
+                if var:
+                    optional_inits.append(If(IfSection(PyccelIsNot(a.var, Nil()),
+                                                       [Assign(var, a.var)])))
+            body.insert2body(*optional_inits, back=False)
+
+            func_kwargs = {
+                    'global_vars':global_vars,
+                    'is_pure':is_pure,
+                    'is_elemental':is_elemental,
+                    'is_private':is_private,
+                    'imports':imports,
+                    'decorators':decorators,
+                    'is_recursive':is_recursive,
+                    'functions': sub_funcs,
+                    'interfaces': func_interfaces,
+                    'result_pointer_map': result_pointer_map,
+                    'docstring': docstring,
+                    'scope': scope,
+            }
+            if is_inline:
+                func_kwargs['namespace_imports'] = namespace_imports
+                global_funcs = [f for f in body.get_attribute_nodes(FunctionDef) if self.scope.find(f.name, 'functions')]
+                func_kwargs['global_funcs'] = global_funcs
+                cls = InlineFunctionDef
+            else:
+                cls = FunctionDef
+            func = cls(name,
+                    arguments,
+                    results,
+                    body,
+                    **func_kwargs)
+            if not is_recursive:
+                recursive_func_obj.invalidate_node()
 
-                results_names = [i.var.name for i in results]
+            if cls_name:
+                # update the class methods
+                if not is_interface:
+                    bound_class.add_new_method(func)
 
-                # Find all nodes which can modify variables
-                assigns = body.get_attribute_nodes(Assign, excluded_nodes = (FunctionCall,))
-                calls   = body.get_attribute_nodes(FunctionCall)
+            funcs += [func]
 
-                # Collect the modified objects
-                lhs_assigns   = [a.lhs for a in assigns]
-                modified_args = [call_arg.value for f in calls
-                                    for call_arg, func_arg in zip(f.args, f.funcdef.arguments) if func_arg.inout]
-                # Collect modified variables
-                all_assigned = [v for a in (lhs_assigns + modified_args) for v in
-                                (a.get_attribute_nodes(Variable) if not isinstance(a, Variable) else [a])]
+        if function_call is not None and len(funcs) == 0:
+            for args in annotated_args:
+                #raise errors if we do not find any compatible function def
+                self._check_argument_compatibility(function_call, args, expr, is_elemental, error=True)
```

Comment:
the annotated_args, are all the arguments of each function in the interface
if `funcs` == 0, it means that we didn't find a compatible function with `functiona_call` arguments

Reply:
So in the case where we cannot find a compatible function the error that is raised is about the passed arguments not matching the types of the first function in the interface? I'm concerned this may be a bit misleading. Couldn't we create a dedicated, more explicit error message?

Reply:
not the first function, if `len(funcs)==0` it means that we have searched all the possibilities and we didn't find a function that matches the call arguments

Reply:
Yes. And then you loop over the `annotated_args`. The first set of `args` corresponds to the first function and `_check_argument_compatibility` will use them to raise the error. So the error that is raised is about the passed arguments not matching the types of the first function in the interface

Reply:
Here is an example reproducing the behaviour I am describing:
```python
def f2():
    print(1)
    return g2(True)

@inline
def g2(a : int | float):
    return 2*a
```
It seems that this only appears in the case of inline functions as for non-inline functions `len(funcs)` is never 0

Reply:
yes, you are right, it does not go over all of them, which probably needs to be done somewhere else, I will remove that line from the method

Reply:
fixed in [a2345c0](https://github.com/pyccel/pyccel/pull/1728/commits/a2345c0e7fb398e7085de2af2084cea811f0d1f9)

Reply:
[8707e5e](https://github.com/pyccel/pyccel/pull/1728/commits/8707e5e515190f20099d19bc59f6ff80b5375b87)